Sources/AsyncHTTPClient/RequestBag+StateMachine.swift

```diff
@@ -0,0 +1,529 @@
+                preconditionFailure("Expected that pause is only called when if we were producing before")
+            case .finished:
+                // the channels writability changed to not writable after we have forwarded the
+                // last bytes from our side.
+                break
+            }
+        case .redirected:
+            // if we are redirected, we should cancel our request body stream anyway
+            break
+        case .finished:
+            // the request is already finished nothing further to do
+            break
+        case .modifying:
+            preconditionFailure("Invalid state")
+        }
+    }
+
+    mutating func resumeRequestBodyStream() -> EventLoopPromise<Void>? {
+        switch self.state {
+        case .initialized, .queued:
+            preconditionFailure("A request stream can only be resumed, if the request was started")
+        case .executing(let executor, let requestState, let responseState):
+            switch requestState {
+            case .initialized:
+                preconditionFailure("Request stream must be started before it can be paused")
+            case .producing:
+                preconditionFailure("Expected that resume is only called when if we were paused before")
+            case .paused(let promise):
+                self.state = .executing(executor, .producing, responseState)
+                return promise
+            case .finished:
+                // the channels writability changed to writable after we have forwarded all the
+                // request bytes. Can be ignored.
+                return nil
+            }
+
+        case .redirected:
+            // if we are redirected, we should cancel our request body stream anyway
+            return nil
+
+        case .finished:
+            preconditionFailure("Invalid state")
+
+        case .modifying:
+            preconditionFailure("Invalid state")
+        }
+    }
+
+    enum WriteAction {
+        case write(IOData, HTTPRequestExecutor, EventLoopFuture<Void>)
+
+        case failTask(Error)
+        case failFuture(Error)
+    }
+
+    mutating func writeNextRequestPart(_ part: IOData, taskEventLoop: EventLoop) -> WriteAction {
+        // this method is invoked with bodyPart and returns a future that signals that
+        // more data can be send.
+        // it may be invoked on any eventLoop
+
+        switch self.state {
+        case .initialized, .queued:
+            preconditionFailure("Invalid state: \(self.state)")
+        case .executing(let executor, let requestState, let responseState):
+            switch requestState {
+            case .initialized:
+                preconditionFailure("Request stream must be started before it can be paused")
+            case .producing:
+                return .write(part, executor, taskEventLoop.makeSucceededFuture(()))
+
+            case .paused(.none):
+                let promise = taskEventLoop.makePromise(of: Void.self)
```

Comment:
This confused me a little until I read the comment in the next case. Worth adding a comment here to say why we're making this promise.